### PR TITLE
[Blazor] Adds support for server side rendered forms (without a body)

### PR DESCRIPTION
**Pull request by javiercn**

### CODE REVIEW

File: src/Components/Components/src/PublicAPI.Unshipped.txt
Comment (javiercn) [inline src/Components/Components/src/PublicAPI.Unshipped.txt:16]: This API was just added for convenience as it cleans up a bunch of cases now that the compiler caches method group to delegate conversions.

File: src/Components/Components/src/PublicAPI.Unshipped.txt
Comment (javiercn) [inline src/Components/Components/src/PublicAPI.Unshipped.txt:54]: This is to avoid allocations altogether if we are not interesting on tracking event handlers. We only track them at the time we want to dispatch an event.

File: src/Components/Components/src/RenderTree/Renderer.cs
Comment (javiercn) [inline src/Components/Components/src/RenderTree/Renderer.cs:390]: We need a new overload to track quiesce when we are dispatching an event to a named event handler. Alternatively, we can make quiescence a feature the renderer offers, and in that way, we don't have to be adding it to every API.

File: src/Components/Components/src/Rendering/ComponentState.cs
Comment (javiercn) [inline src/Components/Components/src/Rendering/ComponentState.cs:79]: We don't do work if you call `SetEventHandlerName` but the renderer is not tracking the event handler names.

File: src/Components/Components/src/Rendering/ComponentState.cs
Comment (javiercn) [inline src/Components/Components/src/Rendering/ComponentState.cs:106]: GetNamedEvents() always returns null unless the renderer is tracking a named event handler.

File: src/Components/Components/src/Rendering/RenderQueueEntry.cs
Comment (javiercn) [inline src/Components/Components/src/Rendering/RenderQueueEntry.cs:23]: I've been adding a bunch of these as I've needed them to make debugging easier

File: src/Components/Components/src/RouteView.cs
Comment (javiercn) [inline src/Components/Components/src/RouteView.cs:86]: Forms bind inside a "context", which has a name. "RouteView" establishes a "default context" matching the path.

File: src/Components/Endpoints/src/RazorComponentEndpointHost.cs
Comment (javiercn) [inline src/Components/Endpoints/src/RazorComponentEndpointHost.cs:42]: This is why we added the `AddComponentParameter` overload

File: src/Components/Endpoints/src/RazorComponentEndpointInvoker.cs
Comment (javiercn) [inline src/Components/Endpoints/src/RazorComponentEndpointInvoker.cs:94]: The "default" handler is the request path, but once we'll support multiple forms, the "handler" query string parameter is what will override the "default" value.

Comment (javiercn) [reply src/Components/Endpoints/src/RazorComponentEndpointInvoker.cs:94]: The "handler" name is consistent with RazorPages

File: src/Components/Components/src/Binding/CascadingModelBinder.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/Binding/CascadingModelBinder.cs:26]: What's the difference between `Name` and `BindingContextId`? Developers might struggle to understand this as they sound like the same thing (also: the XML docs are duplicated).

Is there possibly any other name that could differentiate these more clearly?

File: src/Components/Components/src/Binding/CascadingModelBinder.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/Binding/CascadingModelBinder.cs:42]: I'm not sure this logic will work. If you had a structure like this:

```razor
@page "/mypage"

<CascadingModelBinder>
    <p>The count is @count</p>
</CascadingModelBinder>

<button @onclick="IncrementCounter">Click me</button>
```

... will it behave as expected? I'm suspicious that it will never update the count UI because it's not re-rendering the child content fragment.

File: src/Components/Components/src/Binding/ModelBindingContext.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/Binding/ModelBindingContext.cs:12]: Why use URL.Path and not something like `string.Empty` for the root one? I'm sure you have some reason but it's not obvious to me :)

File: src/Components/Components/src/RenderTree/RenderTreeDiffBuilder.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/RenderTreeDiffBuilder.cs:58]: The comment block seems to make the same point (that enforcement is only during dispatch) three times. Could this be edited down?

File: src/Components/Components/src/RenderTree/RenderTreeDiffBuilder.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/RenderTreeDiffBuilder.cs:27]:
```suggestion
        Dictionary<string, int>? namedEventIndices)
```

Though I feel bad for saying so. If this is not public API, feel free to ignore this point :)

File: src/Components/Components/src/RenderTree/Renderer.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/Renderer.cs:59]:
```suggestion
    }
```

File: src/Components/Components/src/RenderTree/Renderer.cs
Comment (SteveSandersonMS) [reply src/Components/Components/src/RenderTree/Renderer.cs:390]: It's surprising to me that we need to return any quiescence-related `Task` from here. I would have expected that when `Endpoints` dispatches the event, that simply extends the existing `QuiescenceTask` that it was going to wait for anyway.

Perhaps that doesn't work out for some reason that is hard to understand. It would be great to clean this up but I would certainly understand if you don't want to block this PR on it. So I suggest:

 * If you know clearly why it's *not* right for the existing `QuiescenceTask` to be extended by this dispatch, and don't think we should change it to do that, could you say why?
 * Or if you think it would be reasonable for the existing `QuiescenceTask` to be extended by this dispatch, could you file an issue saying we should do that? I'd be happy to dig into what's going on and find out if I can do that.

File: src/Components/Components/src/RenderTree/Renderer.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/Renderer.cs:425]: Doesn't this need to have similar error handling to the branch below? Otherwise it looks like `ErrorBoundary` wouldn't work in this case.

Comment (SteveSandersonMS) [reply src/Components/Components/src/RenderTree/Renderer.cs:425]: We should have an E2E test showing the error handling behavior for form posts (with and without `ErrorBoundary`).

File: src/Components/Web/src/Forms/EditForm.cs
Comment (SteveSandersonMS) [inline src/Components/Web/src/Forms/EditForm.cs:168]: To what extent are you happy with the limitation that SSR event handling only works with `EditForm` and not with a basic `<form>`, even though a basic `<form>` has an `@onsubmit` that looks as if it should also work? This also implies it doesn't work with any custom `MyForm` component that wraps `<form>` without wrapping `EditForm`.

The main alternative I can think of is changing how the named-event tracking works so that:

 * For renderers with some flag enabled (which we only enable for `EndpointHtmlRenderer`), `ComponentState` is updated to track the list of all event names used by a component on its most recent render. There would be no `SetEventHandlerName` API.
 * Then, in `EndpointHtmlRenderer`'s `UpdateDisplayAsync`, it aggregates this to pick out all the `onsubmit` events into a list of component IDs that are currently registering one of those events
 * Then, when trying to dispatch a form post, `EndpointHtmlRenderer` looks at all the components that currently have an `onsubmit` event, and walks through their current rendertree to find the one matching the desired name and its event handler ID

I know the event dispatch phase would involve more processing, but it's hard to see it being problematic because (1) a relatively tiny number of components will register an `onsubmit` event out of the complete set we just rendered, and (2) walking the current rendertree for a component is far less expensive than rendering and diffing it, so this has no impact on the big-O complexity of what we're doing altogether.

This would have the benefit of enabling form-post handling for `<form>` and making it innately work with any `onsubmit` handler that could work during regular interactive rendering.

Is it worth it? I'm open to going either way. I also understand you may not want to make such a big reshuffle before merging this PR. But if you think it's potentially a good improvement for the future, it would be good enough to file an issue and we can cover it in a later preview. I'd be happy to implement this if you want.

File: src/Components/Components/src/RouteView.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/RouteView.cs:20]: Why remove these? Do we now know that the methodgroup-to-delegate conversion is definitely not beneficial to cache?

File: src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs
Comment (SteveSandersonMS) [inline src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs:129]: Despite the comment, I'm still unsure why we need to clear it. Could you update the comment to state what goes wrong if we don't do this? Naively it looks like nothing would be different.

File: src/Components/test/E2ETest/ServerExecutionTests/StreamingRenderingTest.cs
Comment (SteveSandersonMS) [inline src/Components/test/E2ETest/ServerExecutionTests/StreamingRenderingTest.cs:81]: Does this assert pattern really work? It seems like it should at least use `Browser.Equal` instead of `Assert.Equal`, and should do something explicit - such as calling some other API endpoint - to cause the processing to complete otherwise it's a race condition (maybe it will complete before you ever see the "processing" state).

File: src/Components/Components/src/RouteView.cs
Comment (javiercn) [reply src/Components/Components/src/RouteView.cs:20]: https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-11#improved-method-group-conversion-to-delegate

File: src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs
Comment (javiercn) [reply src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs:129]: My goal with clearing it is that we don't have to hold on to the memory, since all these tasks are completed (because we waited for quiesce before dispatching)

File: src/Components/Components/src/RouteView.cs
Comment (SteveSandersonMS) [reply src/Components/Components/src/RouteView.cs:20]: The doc says it is permitted *but not required* to cache the delegate. Does it definitely do so in this case?

File: src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs
Comment (SteveSandersonMS) [reply src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs:129]: OK, still relevant to update the comment, or perhaps better to move the `.Clear` call to immediately after we await them so the association is obvious.

File: src/Components/Components/src/RenderTree/Renderer.cs
Comment (javiercn) [reply src/Components/Components/src/RenderTree/Renderer.cs:390]: @SteveSandersonMS Everything is triggered from
https://github.com/dotnet/aspnetcore/blob/8968058c9e5fdfdd1242426a03dc80609997edab/src/Components/Components/src/RenderTree/Renderer.cs#L229-L238

But when we wait for quiescence, we reset it after we are done (note the finally block):
https://github.com/dotnet/aspnetcore/blob/8968058c9e5fdfdd1242426a03dc80609997edab/src/Components/Components/src/RenderTree/Renderer.cs#L293-L312

File: src/Components/Components/src/Rendering/ComponentState.cs
Comment (javiercn) [reply src/Components/Components/src/Rendering/ComponentState.cs:79]: Actually, I did this for another reason. Tracking event handler names (since we added them to forms) might result in us throwing an exception if two forms end up with the same name, so I want to eliminate any chance that this breaks any existing app that is not ready for handling forms server-side, and I want to eliminate the chance that having duplicate event names precludes the app from rendering, hence why I choose to limit the tracking to when we are trying to dispatch an event (which is the only time when we **really** need to make a decision).

File: src/Components/Components/src/Binding/CascadingModelBinder.cs
Comment (MackinnonBuck) [inline src/Components/Components/src/Binding/CascadingModelBinder.cs:27]: Looks like this comment needs to be updated

File: src/Components/Components/src/RenderTree/RenderTreeDiffBuilder.cs
Comment (javiercn) [reply src/Components/Components/src/RenderTree/RenderTreeDiffBuilder.cs:27]: Is this a UK thing? 😄. I don't mind either way. https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/i/index-indexes-indices

Comment (SteveSandersonMS) [reply src/Components/Components/src/RenderTree/RenderTreeDiffBuilder.cs:27]: It's not important at all! Let's forget about this :)

File: src/Components/Components/src/RouteView.cs
Comment (javiercn) [reply src/Components/Components/src/RouteView.cs:20]: Yes, it is the case since c# 11. https://github.com/dotnet/roslyn/issues/5835

Comment (javiercn) [reply src/Components/Components/src/RouteView.cs:20]: https://github.com/dotnet/roslyn/pull/58288

File: src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs
Comment (javiercn) [reply src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs:129]: Removed the comment and the line, we actually don't need to clear it, since it will be empty anyways, as we only call this after waiting for quiescence on the initial render.

**Comment by javiercn** [comment]
I've done a last round of cleanups and pushed and update. I might add a couple more streaming rendering tests, but otherwise, it should be good to go % feedback.

File: src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs
Comment (MackinnonBuck) [inline src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs:136]:
```suggestion
        // We are generating a path from the root component with component type names like:
```

File: src/Components/test/E2ETest/Microsoft.AspNetCore.Components.E2ETests.csproj
Comment (MackinnonBuck) [inline src/Components/test/E2ETest/Microsoft.AspNetCore.Components.E2ETests.csproj:1]: Are the changes to this file intentional?

**Review by MackinnonBuck — APPROVED** [review]
LGTM, just some small questions/suggestions.

File: src/Components/Endpoints/src/RazorComponentEndpointInvoker.cs
Comment (javiercn) [reply src/Components/Endpoints/src/RazorComponentEndpointInvoker.cs:100]: Yes, this will become async, it's a copy/paste from the full prototype. I will change it when we add antiforgery support.

File: src/Components/Web/src/Forms/EditForm.cs
Comment (javiercn) [reply src/Components/Web/src/Forms/EditForm.cs]: Filed https://github.com/dotnet/aspnetcore/issues/47804 to track revisting this aspect.

File: src/Components/Components/src/RenderTree/Renderer.cs
Comment (javiercn) [reply src/Components/Components/src/RenderTree/Renderer.cs]: Actually, this made me realize that the logic was wrong. We should just throw in this case and let the exception bubble if we are dispatching with quiesce, as the quiesce task will bubble the error.

File: src/Components/test/E2ETest/ServerExecutionTests/StreamingRenderingTest.cs
Comment (javiercn) [reply src/Components/test/E2ETest/ServerExecutionTests/StreamingRenderingTest.cs]: This has been replaced by proper E2E tests

File: src/Components/Components/src/Binding/CascadingModelBinder.cs
Comment (javiercn) [reply src/Components/Components/src/Binding/CascadingModelBinder.cs]: Yeah, I've updated the logic here

File: src/Components/Components/src/Binding/ModelBindingContext.cs
Comment (javiercn) [reply src/Components/Components/src/Binding/ModelBindingContext.cs]: This is using `string.Empty` now

File: src/Components/Components/src/RenderTree/RenderTreeDiffBuilder.cs
Comment (javiercn) [reply src/Components/Components/src/RenderTree/RenderTreeDiffBuilder.cs:51]: Updated the comment

File: src/Components/test/E2ETest/Microsoft.AspNetCore.Components.E2ETests.csproj
Comment (javiercn) [reply src/Components/test/E2ETest/Microsoft.AspNetCore.Components.E2ETests.csproj:1]: No, I've fixed it.

**Pull request merged**
